core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java

```diff
@@ -329,7 +329,13 @@ private void doExecute0(ClientRequestContext ctx, HttpRequestDuplicator rootReqD
         if (!ctx.exchangeType().isResponseStreaming() || config.requiresResponseTrailers()) {
             // XXX(ikhoon): Should we use `response.aggregateWithPooledObjects()`?
             response.aggregate().handle((aggregated, cause) -> {
-                final HttpResponse response0 = cause != null ? HttpResponse.ofFailure(cause) : null;
+                final HttpResponse response0;
+                if (cause != null) {
+                    derivedCtx.logBuilder().endResponse(cause);
```

Comment:
How about passing `cause` directly to `handleResponse` as an argument?
IIUC, it seems possible that `endResponse()` is called first and `.endResponse(cause)` later. If `.endResponse(cause)` is called later, the cause is ignored silently with the current implementation. The cause may happen if there is an async logic with a response in a decorator.
```java
.decorator((delegate, ctx, req) -> {
   final HttpResponse response = delegate.execute(ctx, req);
   return HttpResponse.from(response.aggregate().thenApplyAsync(agg -> {
       return Exceptions.throwUnsafely(new AnticipatedException());
   }));
})
``` 
Or it may happen if we change our implementation unexpectedly by replacing `aggregate().handle()` with `aggregate().handleAsync()`.

Reply:
https://github.com/line/armeria/blob/cae3b6da373101c0976e2d00f1f3cbb706d07e00/core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java#L1352-L1355